src/ring_theory/polynomial/homogeneous.lean

```diff
@@ -225,6 +237,158 @@ noncomputable theory
+and their summations. This matches the form of `mv_polynomial.induction_on'`. -/
+lemma is_homogeneous.induction_on' [comm_semiring R]
+  {p : mv_polynomial σ R → Prop}
+  (hmonomial : ∀ d r, p (monomial d r))
+  (hadd : ∀ j (a b : mv_polynomial σ R),
+    a.is_homogeneous j → b.is_homogeneous j → p a → p b → p (a + b))
+  {x : mv_polynomial σ R} {i : ℕ} (hx : x.is_homogeneous i) : p x :=
+begin
+  have : ∀ x : homogeneous_submodule σ R i, p x :=
+  homogeneous_submodule.induction_on'
+    (λ i, by simpa using hmonomial 0 0)
+    (λ i d r hd, hmonomial _ _)
+    (λ i a b, hadd i _ _ a.prop b.prop),
+  exact this (⟨x, hx⟩ : homogeneous_submodule σ R i)
+end
+
+@[simp]
+lemma to_homogeneous_components_coe [comm_semiring R] {i : ℕ} (x : homogeneous_submodule σ R i) :
+  to_homogeneous_components ↑x = direct_sum.of (λ i, homogeneous_submodule σ R i) i x :=
+begin
+  refine homogeneous_submodule.induction_on' _ _ _ x,
+  { intro i, simp },
+  { intros i d r hd, subst hd, simp, refl },
+  { intros i a b ha hb, simp [ha, hb] },
+end
+
+@[simp]
+lemma to_homogeneous_components_of_is_homogeneous [comm_semiring R] (i : ℕ)
+  (x : mv_polynomial σ R) (hx : x.is_homogeneous i) :
+  to_homogeneous_components x = direct_sum.of (λ i, homogeneous_submodule σ R i) i ⟨x, hx⟩ :=
+(to_homogeneous_components_coe ⟨x, hx⟩ : _)
+
+/-- Assemble a polynomial from a direct sum of homogeneous components. -/
+def of_homogeneous_components [comm_semiring R] :
+  (⨁ i, homogeneous_submodule σ R i) →ₐ[R] mv_polynomial σ R :=
+direct_sum.submodule_coe_alg_hom _
+
+@[simp]
+lemma of_homogeneous_components_of [comm_semiring R] (i : ℕ) (x : homogeneous_submodule σ R i) :
+  of_homogeneous_components (direct_sum.of (λ i, homogeneous_submodule σ R i) i x) = x :=
+direct_sum.to_add_monoid_of _ _ _
+
+/-- `of_*` is the left-inverse of `to_*` -/
+lemma of_to_homogeneous_components [comm_semiring R] :
+  of_homogeneous_components.comp to_homogeneous_components = alg_hom.id R (mv_polynomial σ R) :=
+begin
+  ext : 1,
+  dsimp,
+  rw [to_homogeneous_components_X, of_homogeneous_components_of, subtype.coe_mk],
+end
+
+/-- `of_*` is the left-inverse of `to_*` -/
+lemma to_of_homogeneous_components [comm_semiring R] :
+  to_homogeneous_components.comp of_homogeneous_components =
+    alg_hom.id R (⨁ i, homogeneous_submodule σ R i) :=
+begin
+  ext : 2,
+  dsimp [direct_sum.lof_eq_of],
+  rw [of_homogeneous_components_of, to_homogeneous_components_coe],
+end
+
+/-- Collectively, `mv_polynomial.to_homogeneous_components` and
+`mv_polynomial.of_homogeneous_components` form an equivalence. -/
+@[simps]
+def equiv_homogeneous_components [comm_semiring R] :
+  mv_polynomial σ R ≃ₐ[R] (⨁ i, homogeneous_submodule σ R i) :=
+alg_equiv.of_alg_hom _ _ to_of_homogeneous_components of_to_homogeneous_components
+
+lemma homogeneous_components_is_internal [comm_semiring R] :
+  direct_sum.submodule_is_internal (homogeneous_submodule σ R) :=
```

Comment:
Minor comment: would it be helpful to make the ambient ring explicit? I think this can be a lot more recognizable if it would be written as
```suggestion
  is_graded_algebra (mv_polynomial σ R) (homogeneous_submodule σ R) :=
```
Same definitions, different names and binders.

Reply:
There are two separate things we might want to say about a "decomposition" of a module:

* The decomposition is actually faithful (this is `direct_sum.submodule_is_internal`)
* It preserves multiplication and one in a graded manner (this is `set_like.graded_monoid`)

The first is true and the second is not  in the case where we decompose a module into the submodules corresponding to elements of a basis (with submodules equal to `to_span_singleton.range` or similar).

The second is true and the first is not when the submodules are all just `top`. Maybe that's not so interesting.

Reply:
Aah of course, thanks for clarifying.